tests/test_singularity_fixes.py

```diff
@@ -0,0 +1,107 @@
+import os
+
+from sympy import (
+    Function,
+    Piecewise,
+    exp,
+    log,
+)
+
+from cellmlmanip.model import Quantity, Variable
+from cellmlmanip.parser import Transpiler
+from cellmlmanip.printer import Printer
+
+from . import shared
+
+
+def to_quant(val):
+    return Quantity(val, 'dimensionless')
+
+
+V = Variable(name='V', units='millivolt')
+Z = to_quant(6.2)
+U = to_quant(2) * V + to_quant(5.0)
+SP = (-(log(Z) + to_quant(5.0)) / to_quant(2))
+P = Printer()
+
+
+class exp_(Function):
+
+    def _eval_is_real(self):
+        return self.args[0].is_real
+
+    def fdiff(self, argindex=1):
+        """
+        Returns the first derivative of this function.
+        """
+        assert argindex == 1
+        return self
+
+
+def test_no_V(caplog):
+    model = shared.load_model('test_no_V_no_odes')
+    model.remove_fixable_singularities()
+    assert 'WARNING' in caplog.text
+    assert "Has no membrane_voltage tagged, cannot remove fixable singuarities." in caplog.text
+
+
+def test_fix_singularity_equations():
```

Comment:
I think we should only have one of these "integration tests" on a static model file (ideally one that's in the repo already?)

The rest of the tests should be unit tests where you add a new equation to a model, of some specific form you want to test, and then you just use the method that checks that particular equation and see if it does it correctly.

E.g. in `_get_singularity` there's some complex pattern matching rules (check_bottom_match etc.) so you'd want at least one unit test here per pattern that we match, and then a bunch of tests with slight variations on that pattern that should not be matched, e.g. when you have

```
return m is not None and Z in m and Z != 0 and (
    sp == m[SP_wildcard] or (isinstance(sp, Float)and isinstance(m[SP_wildcard], Float) and isclose(m[SP_wildcard], sp)))
```
I'd expect a test where `m` is None, where Z isn't in m, where Z == 0 etc.

Reply:
fair point I'll change it